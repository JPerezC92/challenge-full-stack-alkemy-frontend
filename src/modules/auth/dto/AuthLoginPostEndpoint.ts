import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";
import { UserEndpoint } from "./UserEndpoint";

export type AuthLoginPostEndpoint =
  | JsendSuccess<{
      accessToken: string;
      refreshToken: string;
      user: UserEndpoint;
    }>
  | JsendError
  | JsendFail;
