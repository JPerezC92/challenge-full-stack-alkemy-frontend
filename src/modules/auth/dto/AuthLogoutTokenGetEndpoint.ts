import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";

export type AuthLogoutTokenGetEndpoint = JsendSuccess | JsendError | JsendFail;
