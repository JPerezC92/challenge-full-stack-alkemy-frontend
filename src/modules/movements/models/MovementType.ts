export enum MovementType {
  INCOME = "income",
  EXPENSE = "expense",
}
