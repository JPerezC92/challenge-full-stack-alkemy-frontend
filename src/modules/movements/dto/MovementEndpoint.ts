import { MovementType } from "src/modules/movements/models/MovementType";

export interface MovementEndpoint {
  id: string;
  concept: string;
  amount: number;
  date: Date;
  type: MovementType;
}
