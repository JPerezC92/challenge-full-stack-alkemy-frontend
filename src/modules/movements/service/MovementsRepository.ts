import { MovementEndpoint } from "src/modules/movements/dto/MovementEndpoint";
import { OrderType } from "src/modules/shared/models/OrderType";
import { MovementCreateDto } from "../dto/MovementCreateDto";
import { MovementType } from "../models/MovementType";

export interface MovementsRepository {
  query(query: {
    limit?: number;
    movementType?: MovementType;
    order?: OrderType;
    page?: number;
  }): Promise<MovementEndpoint[]>;
  create(movementCreateDto: MovementCreateDto): Promise<void>;
}
