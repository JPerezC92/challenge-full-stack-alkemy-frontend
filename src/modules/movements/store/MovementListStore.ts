import { MovementEndpoint } from "../dto/MovementEndpoint";

export interface MovementListStore {
  updateMovementList: (movements: MovementEndpoint[]) => void;
}
