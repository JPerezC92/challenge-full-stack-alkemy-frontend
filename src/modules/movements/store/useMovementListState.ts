import React from "react";
import { MovementEndpoint } from "src/modules/movements/dto/MovementEndpoint";
import { MyStore } from "src/modules/shared/store/MyStore";
import { MovementListStore } from "./MovementListStore";

export function useMovementListState() {
  const [movementList, setMovementList] = React.useState<MovementEndpoint[]>(
    []
  );

  const movementListStore: MyStore<MovementListStore> = React.useCallback(
    () => ({
      updateMovementList: (movements: MovementEndpoint[]) =>
        setMovementList(movements),
    }),
    []
  );

  return { movementList, movementListStore };
}
