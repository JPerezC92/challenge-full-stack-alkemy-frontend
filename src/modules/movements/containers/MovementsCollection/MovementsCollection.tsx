import { MovementCard } from "src/modules/movements/components/MovementCard";
import { MovementType } from "src/modules/movements/models/MovementType";
import { MovementsRepository } from "src/modules/movements/service/MovementsRepository";
import { useMovementListState } from "src/modules/movements/store/useMovementListState";
import { useRequest } from "src/modules/shared/hooks/useRequest";
import { OrderType } from "src/modules/shared/models/OrderType";
import { MyRepository } from "src/modules/shared/service/MyRepository";

type MovementsCollectionProps = {
  movementsRepository: MyRepository<MovementsRepository>;
  movementType: MovementType;
};

export const MovementsCollection: React.FC<MovementsCollectionProps> = ({
  movementsRepository,
  movementType,
}) => {
  const { movementList, movementListStore } = useMovementListState();

  useRequest(
    async ({ abortController }) => {
      const _movementsRepository = movementsRepository({ abortController });
      const _movementListStore = movementListStore();
      const movementList = await _movementsRepository.query({
        movementType,
        order: OrderType.DESC,
      });

      _movementListStore.updateMovementList(movementList);
    },
    [movementListStore, movementType, movementsRepository]
  );

  return (
    <div>
      {movementList.map((movement) => (
        <MovementCard key={movement.id} {...movement} />
      ))}
    </div>
  );
};
