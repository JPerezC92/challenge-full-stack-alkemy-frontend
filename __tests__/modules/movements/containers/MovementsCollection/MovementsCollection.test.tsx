import { render, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import { MovementsCollection } from "src/modules/movements/containers/MovementsCollection";
import { MovementEndpoint } from "src/modules/movements/dto/MovementEndpoint";
import { MovementType } from "src/modules/movements/models/MovementType";
import { MovementsRepository } from "src/modules/movements/service/MovementsRepository";
import * as useNodeMovementsRepository from "src/modules/movements/service/useNodeMovements.repository";
import { MovementListStore } from "src/modules/movements/store/MovementListStore";
import * as useMovementListState from "src/modules/movements/store/useMovementListState";
import { OrderType } from "src/modules/shared/models/OrderType";

const movementList: MovementEndpoint[] = Array.from({ length: 10 }).map(
  (_, index) => ({
    id: (index + 1).toString(),
    type: MovementType.INCOME,
    amount: 1000 * (index + 1),
    concept: `concept ${index + 1}`,
    date: `2020-01-${index + 1}`,
  })
);

const movementsRepository: MovementsRepository = {
  query: jest.fn().mockResolvedValue(movementList),
  create: jest.fn(),
};

const movementListStore: MovementListStore = {
  updateMovementList: jest.fn(),
};

jest
  .spyOn(useNodeMovementsRepository, "useNodeMovementsRepository")
  .mockReturnValue(() => movementsRepository);

jest.spyOn(useMovementListState, "useMovementListState").mockReturnValue({
  movementList,
  movementListStore: () => movementListStore,
});

describe("MovementsCollection container", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("should match the snapshot", async () => {
    const component = renderer
      .create(
        <MovementsCollection
          movementType={MovementType.INCOME}
          movementsRepository={useNodeMovementsRepository.useNodeMovementsRepository()}
        />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  test("should first", async () => {
    const component = render(
      <MovementsCollection
        movementType={MovementType.INCOME}
        movementsRepository={useNodeMovementsRepository.useNodeMovementsRepository()}
      />
    );

    await waitFor(async () =>
      expect(await component.findAllByRole("article")).toHaveLength(
        movementList.length
      )
    );

    expect(movementsRepository.query).toHaveBeenCalledTimes(1);
    expect(movementsRepository.query).toHaveBeenCalledWith({
      movementType: MovementType.INCOME,
      order: OrderType.DESC,
    });
    expect(movementListStore.updateMovementList).toHaveBeenCalledTimes(1);
    expect(movementListStore.updateMovementList).toHaveBeenCalledWith(
      movementList
    );
  });
});
