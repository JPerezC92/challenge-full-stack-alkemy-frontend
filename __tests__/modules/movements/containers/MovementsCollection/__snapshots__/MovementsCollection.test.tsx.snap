// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovementsCollection container should match the snapshot 1`] = `
<div>
  <article>
    <p>
      1
    </p>
    <p>
      1000
    </p>
    <p>
      concept 1
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      2
    </p>
    <p>
      2000
    </p>
    <p>
      concept 2
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      3
    </p>
    <p>
      3000
    </p>
    <p>
      concept 3
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      4
    </p>
    <p>
      4000
    </p>
    <p>
      concept 4
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      5
    </p>
    <p>
      5000
    </p>
    <p>
      concept 5
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      6
    </p>
    <p>
      6000
    </p>
    <p>
      concept 6
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      7
    </p>
    <p>
      7000
    </p>
    <p>
      concept 7
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      8
    </p>
    <p>
      8000
    </p>
    <p>
      concept 8
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      9
    </p>
    <p>
      9000
    </p>
    <p>
      concept 9
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
  <article>
    <p>
      10
    </p>
    <p>
      10000
    </p>
    <p>
      concept 10
    </p>
    <p>
      income
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </article>
</div>
`;
